runtime/runtime_controller.cc

```diff
@@ -50,48 +53,76 @@ RuntimeController::RuntimeController(
-    root_isolate_return_code_ = {true, code};
-  });
-
-  if (auto* platform_configuration = GetPlatformConfigurationIfAvailable()) {
-    tonic::DartState::Scope scope(strong_root_isolate);
-    platform_configuration->DidCreateIsolate();
-    if (!FlushRuntimeStateToIsolate()) {
-      FML_DLOG(ERROR) << "Could not setup initial isolate state.";
-    }
-  } else {
-    FML_DCHECK(false) << "RuntimeController created without window binding.";
-  }
-
-  FML_DCHECK(Dart_CurrentIsolate() == nullptr);
+  create_and_config_root_isolate_ =
+      std::async(std::launch::deferred, [self = weak_factory_.GetWeakPtr()]() {
+        if (!self) {
+          return;
+        }
+
+        auto strong_root_isolate =
+            DartIsolate::CreateRootIsolate(
+                self->vm_->GetVMData()->GetSettings(),                //
+                self->isolate_snapshot_,                              //
+                self->task_runners_,                                  //
+                std::make_unique<PlatformConfiguration>(self.get()),  //
+                self->snapshot_delegate_,                             //
+                self->io_manager_,                                    //
+                self->unref_queue_,                                   //
+                self->image_decoder_,                                 //
+                self->advisory_script_uri_,                           //
+                self->advisory_script_entrypoint_,                    //
+                nullptr,                                              //
+                self->isolate_create_callback_,                       //
+                self->isolate_shutdown_callback_                      //
+                )
+                .lock();
+
+        FML_CHECK(strong_root_isolate) << "Could not create root isolate.";
+
+        // The root isolate ivar is weak.
+        self->root_isolate_ = strong_root_isolate;
+
+        strong_root_isolate->SetReturnCodeCallback([self](uint32_t code) {
+          if (!self) {
+            return;
+          }
+
+          self->root_isolate_return_code_ = {true, code};
+        });
+
+        if (auto* platform_configuration =
+                self->GetPlatformConfigurationIfAvailable()) {
+          tonic::DartState::Scope scope(strong_root_isolate);
+          platform_configuration->DidCreateIsolate();
+          if (!self->FlushRuntimeStateToIsolate()) {
+            FML_DLOG(ERROR) << "Could not setup initial isolate state.";
+          }
+        } else {
+          FML_DCHECK(false)
+              << "RuntimeController created without window binding.";
+        }
+
+        FML_DCHECK(Dart_CurrentIsolate() == nullptr);
+
+        self->client_.OnRootIsolateCreated();
+        return;
+      });
+
+  task_runners_.GetUITaskRunner()->PostTask(
+      [self = weak_factory_.GetWeakPtr()]() {
+        if (!self) {
+          return;
+        }
+
+        self->GetRootIsolate();
```

Comment:
Nit: add a comment here that we're still trying to create the root isolate as soon as possible here on the UI thread although it's deferred a little bit by `std::async(std::launch::deferred, ...)`. So the callers of `GetRootIsolate` should get a quick return after this UI thread task.

Reply:
Done